[REST | SPARK]: Send referenced-by to all endpoints

singhpk234 opened the pull request:
### About the change 

This provides a reference implementation for passing the view name that table is referenced in as part of which the loadTable call is being made. 

Details on the spec change proposal and motivation here 
- https://github.com/apache/iceberg/pull/13810

gaborkaszab reviewed (commented):
Thanks for the PR @singhpk234 !
I took a look for my own benefit, and left some observations and questions meanwhile. Let me know what you think!

gaborkaszab commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -367,46 +368,67 @@ public boolean tableExists(SessionContext context, TableIdentifier identifier) {
+    Map<String, String> queryParams = Maps.newHashMap(mode.params());
+
+    Object viewIdentifierObj = viewContext.get(ContextAwareTableCatalog.VIEW_IDENTIFIER_KEY);
+    if (viewIdentifierObj instanceof String) {
```
If there is a view identifier provided here but it's not string, or it has less than 2 parts we silently omit the provided input. Shouldn't we raise an illegal argument exception rather to tell the clients that whatever they provided was incorrect?

nandorKollar replied:
+1

gaborkaszab commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -367,46 +368,67 @@ public boolean tableExists(SessionContext context, TableIdentifier identifier) {
+        String[] namespaceParts = new String[parts.length - 1];
+        System.arraycopy(parts, 0, namespaceParts, 0, parts.length - 1);
+        String viewName = parts[parts.length - 1];
+        queryParams.put(
```
Just for my understanding:
Let's take the following input: "ns1.ns2.view1"
We convert this to "ns1%1Fns2%1F.view1"
Is this because we wan't to tell the REST catalog which part is the namespace and which is the view name? If we include a dot in the query params, can't we just send the view path as it is with dots and let the catalog server figure out the namespace/view name part that we do here? Or can we also use the same special char between the namespace and the view name?

singhpk234 replied:
one way to look at this is anything from the last `.` to the end is the view name and the every thing before it is the namespace, the reason `%1F` was chosen the separator was to use something similar, the reason i choose the `%1F` is the language of the existing rest spec for the nested namespaces.

I think we are yet to conclude on the separator discussion.

singhpk234 commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:
```diff
@@ -167,16 +167,28 @@ protected TableIdentifier buildIdentifier(Identifier identifier) {
+  }
+
+  @Override
+  public Table loadTableViaView(Identifier ident, String version, Map<String, Object> context)
```
Renamed it to loadTableWithContext where context would be a runtime context, let me know does is sounds better !

singhpk234 commented on api/src/main/java/org/apache/iceberg/catalog/ContextAwareTableCatalog.java:
```diff
@@ -0,0 +1,51 @@
+public interface ContextAwareTableCatalog {
+
+  /** Context key for the view identifier that references a table */
+  String VIEW_IDENTIFIER_KEY = "view.identifier";
```
makes sense ! i kept the `view.` prefix though !

nastra commented on api/src/main/java/org/apache/iceberg/catalog/ContextAwareTableCatalog.java:
```diff
@@ -0,0 +1,51 @@
+   * @return the loaded table
+   * @throws NoSuchTableException if the table does not exist
+   */
+  Table loadTableWithContext(TableIdentifier identifier, Map<String, Object> context)
```
what about just calling this `loadTable` as the "context" part is already implicitly included in the paramters

nastra commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/views/UnResolvedRelationFromView.scala:
```diff
@@ -0,0 +1,66 @@
+ * - Definer rights: Use view creator's permissions to access the underlying table
+ * - Invoker rights: Use current user's permissions to access the underlying table
+ *
+ * When passed to REST catalogs, the view identifier is encoded using unit separator
```
I don't think this should have any catalog-specific info in it

singhpk234 replied:
Agree, this is mostly done to show E2E use case, ideally any catalog which implements the ContextAwareTableCatalog impl should be able to leverage this, let me remove this line

gaborkaszab commented on api/src/main/java/org/apache/iceberg/catalog/ContextAwareTableCatalog.java:
```diff
@@ -0,0 +1,62 @@
+   * <p>Common context keys:
+   *
+   * <ul>
+   *   <li>{@link #VIEW_IDENTIFIER_KEY}: A List&lt;TableIdentifier&gt; of view(s) referencing this
```
Is it necessary to escape the lt and gt chars?
"List&lt;TableIdentifier&gt;"

gaborkaszab commented on api/src/main/java/org/apache/iceberg/catalog/SessionCatalog.java:
```diff
@@ -361,4 +361,9 @@ default boolean namespaceExists(SessionContext context, Namespace namespace) {
+
+  default Table loadTableWithContext(
+      SessionContext sessionContext, TableIdentifier ident, Map<String, Object> context) {
+    return loadTable(sessionContext, ident);
```
Have you considered throwing UnsupportedOperationException instead of silently ignore the context param?

gaborkaszab commented on core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java:
```diff
@@ -159,5 +159,10 @@ public boolean removeProperties(Namespace namespace, Set<String> removals) {
+
+    @Override
+    public Table loadTable(TableIdentifier identifier, Map<String, Object> loadingContext) {
+      return BaseSessionCatalog.this.loadTableWithContext(context, identifier, loadingContext);
```
nit: maybe consider renaming `context` to `sessionContext` to avoid confusion with the new type of context?

gaborkaszab commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -396,46 +397,96 @@ private static Map<String, String> snapshotModeToParam(SnapshotMode mode) {
+    Map<String, String> queryParams = Maps.newHashMap(params);
+    Object viewIdentifierObj = context.get(ContextAwareTableCatalog.VIEW_IDENTIFIER_KEY);
+
+    if (!(viewIdentifierObj instanceof List)) {
```
nit: maybe `Preconditions.checkArgument(viewIdentifierObj instanceof List, error msg)`

gaborkaszab commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -396,46 +397,96 @@ private static Map<String, String> snapshotModeToParam(SnapshotMode mode) {
+
   @Override
-  public Table loadTable(SessionContext context, TableIdentifier identifier) {
+  public Table loadTableWithContext(
```
Maybe this had been discussed already, but can't this be simply `loadTable`? The params will tell us anyway that this is with context.

gaborkaszab reviewed (commented):
I went through the api/ and core/ part for now (haven't checked Spark). Left some questions and nits. I also found some of my previous comments so I'm not sure this was ready for review.
Anyway, some test coverage in TestRESTCatalog would be great!

sfc-gh-prsingh committed:
Add E2E with all endpoints with the approved spec version

singhpk234 commented:
Thanks @gaborkaszab i marked this as draft to clean this more up, specially now since the spec is in, I can iron this more. will re-request review from everyone soon !

sfc-gh-prsingh committed:
plumb view

sfc-gh-prsingh committed:
add more view e2e plumbing

sfc-gh-prsingh committed:
Rename ContextAwareTableCatalog to ContextAwareCatalog and clean up

- Rename interface to reflect it handles both tables and views
- Add loadView as default method on ContextAwareCatalog (API module)
- Remove loadView from ViewCatalog (now on ContextAwareCatalog)
- Extract extractViewChain helper in RESTSessionCatalog
- Remove dangerous catch-all in ResolveViews.scala
- Fix inline import in ResolveViews.scala
- Fail explicitly when catalog doesn't support context-aware loading

sfc-gh-prsingh committed:
make ci happy

sfc-gh-prsingh committed:
claenup more

gaborkaszab commented on api/src/main/java/org/apache/iceberg/catalog/ContextAwareCatalog.java:
```diff
@@ -0,0 +1,78 @@
+   * @throws NoSuchTableException if the table does not exist
+   */
+  default Table loadTable(TableIdentifier identifier, Map<String, Object> loadingContext)
+      throws NoSuchTableException {
```
Probably no need to add this here. I don't see other interfaces having 'throws NoSuchTableException'

gaborkaszab reviewed (commented):
Thank you for the PR @singhpk234 !
I went through the non-Spark part. Let me know what you think!

bryanck commented on spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/SparkSessionCatalog.java:
```diff
@@ -143,7 +146,22 @@ public Identifier[] listTables(String[] namespace) throws NoSuchNamespaceExcepti
+  public Table loadTable(Identifier ident, Map<String, Object> context)
+      throws NoSuchTableException {
     try {
+      if (!context.isEmpty() && icebergCatalog instanceof ContextAwareCatalog) {
```
When `SparkCatalog` is initialized, it will wrap the Iceberg catalog in a `CachingCatalog` by default, unless caching is disabled, so context won't get passed in that case. Does it make sense to implement `ContextAwareCatalog` in `CachingCatalog` also?

bryanck commented on spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/ContextAwareCatalog.java:
```diff
@@ -0,0 +1,80 @@
+ * context. The loading context can carry information such as the chain of views that reference a
+ * table, enabling catalogs to make context-aware decisions (e.g., scoped credential vending).
+ */
+public interface ContextAwareCatalog {
```
I feel like we might want to rename this for clarity, as it conflicts with `org.apache.iceberg.catalog.ContextAwareCatalog`

sfc-gh-prsingh committed:
first feedback

ldsantos0911 commented on spark/v4.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveViews.scala:
```diff
@@ -161,17 +220,29 @@ case class ResolveViews(spark: SparkSession) extends Rule[LogicalPlan] with Look
+        viewChain match {
+          case Some(chain) =>
+            UnResolvedRelationFromView(qualifiedTableId, chain, options, isStreaming)
+          case _ =>
```
This should be impossible, shouldn't it? Would it make sense to throw an error in this case?

ldsantos0911 commented on spark/v4.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveViews.scala:
```diff
@@ -62,6 +68,48 @@ case class ResolveViews(spark: SparkSession) extends Rule[LogicalPlan] with Look
+          ViewUtil
+            .loadView(catalog, tableIdent, context)
+            .map(view => createViewRelation(tableParts, view, Some(viewChain)))
+            .getOrElse(UnresolvedRelation(tableParts, options, isStreaming))
```
Similar to below, do we want to be falling back to the default resolution path, or would it make sense to throw an error if the expected view chain is broken somehow?

ldsantos0911 commented on spark/v4.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/views/UnResolvedRelationFromView.scala:
```diff
@@ -0,0 +1,67 @@
+
+  override def toString: String = {
+    val chainStr = viewChain.map(parts => s"[${parts.mkString(", ")}]").mkString(", ")
+    s"UnresolvedRelationFromView([${tableMultipartIdentifier.mkString(", ")}], " +
```
nit: The capitalization is mismatched between the string representation and the class name. My vote would be to align with the capitalization in this method.

sfc-gh-prsingh committed:
address PR feedback

- Rename ContextAwareTableCatalog to SparkContextAwareCatalog (bryanck)
- Add spark.sql.iceberg.referenced-by-enabled session config to gate feature
- Add strict cross-catalog validation with IllegalStateException
- Qualify view chain entries as fully-qualified identifiers
- Handle RelationTimeTravel in view chain resolution
- Fix capitalization in UnResolvedRelationFromView.simpleString
- Separate referenced-by constants from snapshot block in RESTCatalogProperties
- Add E2E tests for loadTable with referenced-by query params

sfc-gh-prsingh committed:
cleanup javadocs, dedup test setup and identifier qualification

sfc-gh-prsingh committed:
remove default impls from SessionCatalog and ViewSessionCatalog, extract buildViewChain to ViewUtil

sfc-gh-prsingh committed:
move extractViewChain and referencedByToQueryParam to RESTUtil

sfc-gh-prsingh committed:
restore default impls on SessionCatalog and ViewSessionCatalog to avoid API break

singhpk234 commented on api/src/main/java/org/apache/iceberg/catalog/ContextAwareCatalog.java:
```diff
@@ -0,0 +1,60 @@
+   * @return instance of {@link Table} referred by the identifier
+   * @throws UnsupportedOperationException if context-aware loading is not supported
+   */
+  default Table loadTable(TableIdentifier identifier, Map<String, Object> loadingContext) {
```
Ack on this ! How about 
``` java
loadTable(TableIdentifier identifier, List<TableIdentifier> refChain)
```
as an API

sfc-gh-prsingh committed:
refactor per new api proposal

bryanck commented on spark/v4.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveViews.scala:
```diff
@@ -62,6 +67,33 @@ case class ResolveViews(spark: SparkSession) extends Rule[LogicalPlan] with Look
+          isStreaming,
+          timeTravelVersion,
+          timeTravelTimestamp) =>
+      val referencedBy = ViewUtil.buildReferencedByChain(viewChain, catalog.name())
```
I'm wondering if it's feasible to use the view UUID as the referenced by instead of the view name identifier? That might address issues around cross catalog access and having the catalog name in the identifier.

singhpk234 replied:
I gave it more thought, the thing is catalog don't index things from uuid, all our specification is based on identifiers / name, hence this would be a tricky, do you have any thoughts on server side implementation pov ?

RjLi13 commented on core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:
```diff
@@ -386,4 +390,53 @@ public static Map<String, String> configHeaders(Map<String, String> properties)
+            .map(
+                ident ->
+                    encodeNamespace(ident.namespace(), namespaceSeparator)
+                        + namespaceSeparator
```
Should the namespaceSeparator be url decoded first like in namespaceToQueryParam? Otherwise server would receive prod%251F%251Fanalytics%251Fquarterly_view with %1F being url-encoded again.

sfc-gh-prsingh committed:
match referenced-by spec wire form and drop sql feature flag

- emit referenced-by in wire form (URL-encoded levels/name joined by
  the URL-encoded separator, entries by literal comma)
- bypass URIBuilder.addParameter for referenced-by so the comma chain
  delimiter survives unchanged on the wire (matches openapi example)
- always send referenced-by when resolving inside a view body, since
  opt-in is now an api-level construct (the SupportsReferencedBy
  interface); the spark.sql.iceberg.referenced-by-enabled flag is
  redundant and is removed

sfc-gh-prsingh committed:
extract ViewUtil.qualifyParts and trim comments

- share qualifyParts between buildViewChain and qualifyTableIdentifiers
  instead of duplicating the same Seq[String] match in both files
- trim HTTPRequest.java referenced-by comment to the WHY only
- drop the "...sent verbatim on the wire" trailers in test comments

RjLi13 commented on spark/v4.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveViews.scala:
```diff
@@ -102,12 +131,19 @@ case class ResolveViews(spark: SparkSession) extends Rule[LogicalPlan] with Look
-    val rewritten = rewriteIdentifiers(parsed, viewCatalogAndNamespace);
+
+    val viewChain =
+      ViewUtil.buildViewChain(nameParts, viewCatalogAndNamespace, existingChain, isCatalog)
```
If existingChain is empty, should the viewChain be built initially with the current session's catalog and namespace instead of `viewCatalogAndNamespace` at time of creation? For example, the view was created in a session where default namespace is `default` but qualified as `create view analytics.view ...`. Then, the user in a new session does this sql query 
```sql
USE analytics;
select * from view
```
the referenced-by would contain [default.view] instead of [analytics.view] (applicable for outermost view only). This seems to mirror the load functionality, using current session catalog at query time before using the view catalog and namespace for the body.

ldsantos0911 replied:
+1

singhpk234 replied:
Good catch! You're right `viewCatalogAndNamespace` (the spec's default-catalog/default-namespace for view) represents the session context at view creation time, meant only for resolving unqualified references inside the view's SQL body. It shouldn't be used to identify the view itself in the referenced-by chain.

Fixed by using the session-resolved identity from CatalogAndIdentifier for the outermost view's chain entry. viewCatalogAndNamespace is still used only for rewriteIdentifiers (its correct purpose per spec).
